tests/Frontend/CoreSpec.hs

```diff
@@ -263,8 +263,10 @@ postToForm (F g) = do
         let frm = makeForm page
         env <- runFailOnError $ (`payloadToEnv` payload) <$> getForm "" frm
 
-        (_, Just payload') <- runFailOnError $ postForm "" frm (\_ -> pure env)
-        liftIO $ payload' `shouldBe` payload
+        (_, mpayload) <- runFailOnError $ postForm "" frm (\_ -> pure env)
+        case mpayload of
+            Nothing       -> fail "Form validation has failed."
+            Just payload' -> liftIO $ payload' `shouldBe` payload
```

Comment:
It would be even nicer if we could somehow get the validation errors here.  Just an idea, feel free to ignore.